pkg/controller/mattermostcluster/mattermostcluster_controller.go

```diff
@@ -100,54 +110,524 @@ func (r *ReconcileMattermostCluster) Reconcile(request reconcile.Request) (recon
+					Version: mattermostv1alpha1.SchemeGroupVersion.Version,
+					Kind:    "MattermostCluster",
+				}),
+			},
+			Annotations: map[string]string{
+				"kubernetes.io/ingress.class": "nginx",
+				//"kubernetes.io/tls-acme":      "true",
+			},
+		},
+		Spec: spec,
+	}
+
+	foundIngress := &v1beta1.Ingress{}
+	errGet := r.client.Get(context.TODO(), types.NamespacedName{Name: ingressName, Namespace: mattermost.Namespace}, foundIngress)
+	if errGet != nil && errors.IsNotFound(errGet) {
+		reqLogger.Info("MattermostCluster Ingress not found, will deploy one.", "Err", errGet.Error())
+		reqLogger.Info("Creating Mattermost Ingress")
+		if err := r.client.Create(context.TODO(), ingressMM); err != nil {
+			reqLogger.Info("Error creating Ingress", "Error", err.Error())
+			return err
+		}
+		reqLogger.Info("Creating Mattermost Ingress completed")
+		if err := controllerutil.SetControllerReference(mattermost, ingressMM, r.scheme); err != nil {
+			return err
+		}
+	} else if errGet != nil {
+		reqLogger.Error(errGet, "MattermostCluster Ingress")
+		return errGet
+	}
+
+	// TODO check how to do the update
+	// if !reflect.DeepEqual(ingressMM, foundIngress) {
+	// 	foundIngress = ingressMM
+	// 	reqLogger.Info("Updating Mattermost Ingress")
+	// 	err = r.client.Update(context.TODO(), foundIngress)
+	// 	if err != nil {
+	// 		return err
+	// 	}
+	// _ = controllerutil.SetControllerReference(mattermost, foundIngress, r.scheme)
+	// }
+	return nil
+}
+
+func (r *ReconcileMattermostCluster) deploymentForMM(mattermost *mattermostv1alpha1.MattermostCluster, reqLogger logr.Logger) error {
+	dbPassword, err := r.getSecrets(mattermost, reqLogger)
+	if err != nil {
+		return fmt.Errorf("Error getting the database password. Err=%s", err.Error())
+	}
+	datasourceMM := fmt.Sprintf("mysql://root:%s@tcp(%s-db.%s:3306)/mattermost?charset=utf8mb4,utf8&readTimeout=30s&writeTimeout=30s", dbPassword, mattermost.Name, mattermost.Namespace)
+	initContainerCMD := fmt.Sprintf("until curl --max-time 5 http://%s-db.%s:3306; do echo waiting for mysql; sleep 5; done;", mattermost.Name, mattermost.Namespace)
+	cmdInit := []string{"sh", "-c"}
+	cmdInit = append(cmdInit, initContainerCMD)
+	cmdStartMM := []string{"mattermost"} //, "--config", datasourceMM}
+
+	cmdInitDB := []string{"sh", "-c"}
+	cmdInitDB = append(cmdInitDB, fmt.Sprintf("mysql -h %s-db.%s -u root -p%s -e 'CREATE DATABASE IF NOT EXISTS mattermost'", mattermost.Name, mattermost.Namespace, dbPassword))
+
+	deployMM := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      mattermost.Name,
+			Namespace: mattermost.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(mattermost, schema.GroupVersionKind{
+					Group:   mattermostv1alpha1.SchemeGroupVersion.Group,
+					Version: mattermostv1alpha1.SchemeGroupVersion.Version,
+					Kind:    "MattermostCluster",
+				}),
+			},
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &mattermost.Spec.Replicas,
```

Comment:
One thing that's not clear to me is how the reconciler responds to changes in data, vs. just creating them: say if the `Replicas` value increases -- it /seems/ like this code can only handle creating the deployment, and not modify the existing one.

Or can you elaborate how this works?

Reply:
explained in the call